### PR TITLE
New Tool: `update-go-azure-sdk`

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR introduces a new tool `update-go-azure-sdk` which is primarily intended to be run in automation.

This tool obtains the changes between two versions of `hashicorp/go-azure-sdk`, then updates the version of `hashicorp/go-azure-sdk` used in this repository - and when a new API Version is available in the latest SDK, attempts to update the API Version for each Service.

Both the SDK Upgrade and each API Version update are done in individual commits - and prior to both being committed the unit tests are run to ensure these pass. This means that when an API Version is compatible, that is contains no breaking changes to the API Payloads - the API Version upgrade can be done automatically. However in the event that the API Payload is _not_ compatible - the API Version will be ignored, with the issues outlined in the PR description. Examples of both of these can be seen below.

This will be conditionally run by `hashicorp/go-azure-sdk` after a release is made, to both update the version of `hashicorp/go-azure-sdk` that's being used in the Provider - but also to update the API Version being used in the Provider when a new API Version becomes available (and the build passes).

There's a bunch of assumptions built into this tool and it's not going to be infallible - however this should both ensure that `hashicorp/go-azure-sdk` is up-to-date in this repository and remove the manual work needed to update from one API Version to another.

## Examples

Console output:

```bash$ go build . && ./update-go-azure-sdk --old-sdk-version=v0.20230918.1115907 --new-sdk-version=v0.20231005.1153009 --working-directory=../../../ --output-file=pr-description.txt
2023-10-10T15:14:12.249+0200 [INFO]  Old SDK Version is "v0.20230918.1115907"
2023-10-10T15:14:12.249+0200 [INFO]  New SDK Version is "v0.20231005.1153009"
2023-10-10T15:14:12.249+0200 [INFO]  Output File Name is "pr-description.txt"
2023-10-10T15:14:12.249+0200 [INFO]  Working Directory is "../../../"
2023-10-10T15:14:12.249+0200 [INFO]  Checking the changes between "v0.20230918.1115907" and "v0.20231005.1153009" of `hashicorp/go-azure-sdk`..
2023-10-10T15:14:34.275+0200 [INFO]  Updating `hashicorp/go-azure-sdk`..
2023-10-10T15:14:50.939+0200 [INFO]  Committed as "b5b40b0b49d9f192f9d65d620ecba98862764e6b"
2023-10-10T15:14:53.958+0200 [INFO]  Processing Service "connectedvmware"..
2023-10-10T15:14:54.552+0200 [INFO]  Processed Service "connectedvmware".
2023-10-10T15:14:57.578+0200 [INFO]  Processing Service "cosmosdb"..
2023-10-10T15:14:58.148+0200 [INFO]  Attempting to update API Version "2022-05-15" to "2023-09-15" for Service "cosmosdb"..
2023-10-10T15:15:03.823+0200 [INFO]  Updated the Imports for Service "cosmosdb" to use API Version "2023-09-15" rather than "2022-05-15"..
2023-10-10T15:15:03.823+0200 [INFO]  Running `go mod tidy` / `go mod vendor`..
2023-10-10T15:15:15.022+0200 [INFO]  Running `make test` within "../../../"..
2023-10-10T15:18:04.450+0200 [INFO]  Committed as "a88ab1b1be8344a07bcf316a62c3ea31a87be507"
2023-10-10T15:18:04.450+0200 [INFO]  Updated Service "cosmosdb" from "2022-05-15" to "2023-09-15"
2023-10-10T15:18:04.450+0200 [INFO]  Attempting to update API Version "2022-11-15" to "2023-09-15" for Service "cosmosdb"..
2023-10-10T15:18:10.035+0200 [INFO]  Updated the Imports for Service "cosmosdb" to use API Version "2023-09-15" rather than "2022-11-15"..
2023-10-10T15:18:10.035+0200 [INFO]  Running `go mod tidy` / `go mod vendor`..
2023-10-10T15:18:22.087+0200 [INFO]  Running `make test` within "../../../"..
2023-10-10T15:19:46.236+0200 [INFO]  Resetting the working directory since `make test` failed..
2023-10-10T15:19:56.549+0200 [INFO]  Attempting to update API Version "2023-04-15" to "2023-09-15" for Service "cosmosdb"..
2023-10-10T15:20:02.211+0200 [INFO]  Updated the Imports for Service "cosmosdb" to use API Version "2023-09-15" rather than "2023-04-15"..
2023-10-10T15:20:02.211+0200 [INFO]  Running `go mod tidy` / `go mod vendor`..
2023-10-10T15:20:12.584+0200 [INFO]  Running `make test` within "../../../"..
```

Output in a PR:

Screenshot 2023-10-10 at 12 54 56

An earlier build log containing example failures (with slightly less detail) can be seen here: https://gist.github.com/tombuildsstuff/d6c153d484697961b6519a3c1cec4882#file-output-txt-L156-L337

### Why not extend the existing tool `update-api-version`?

In short, there's value in both - updating the API Version being used, and updating `go-azure-sdk` are subtly different use-cases - with the former being primarily run by a human and the latter  primarily being run in automation.

It's worth calling out that some of the changes made in this PR should also be made to that tool, but that'll follow as a separate PR.

### Why not just use Dependabot?

Dependabot would work great for updating the SDK dependency - however this wouldn't allow us to update the API Versions being used in the Provider. In addition, using a custom tool allows us to provide additional logic, such as running the (unit) tests for each API Version update to ensure these pass prior to committing the changes - else rolling this back (and documenting it in the PR description).

Whilst there's value in running this on every release of `hashicorp/go-azure-sdk` - for the moment we need this tool to be conditionally run, as when the base layer is migrated from `Azure/go-autorest` to `hashicorp/go-azure-sdk` additional changes must be made by hand.

### Why not use a bash script?

In short, it's far simpler to get the output we want from a tool.

### CODE REVIEW

File: internal/tools/update-go-azure-sdk/main.go [hunk 1]
```diff
@@ -0,0 +1,729 @@
...
+		"mod",
+		"tidy",
+	}
+	tidyCmd := exec.CommandContext(ctx, "go", tidyArgs...)
+	tidyCmd.Dir = workingDirectory
+	_ = tidyCmd.Start()
+	_ = tidyCmd.Wait()
+	logger.Debug(fmt.Sprintf("Run `go mod tidy` in %q.", workingDirectory))
+
+	logger.Debug(fmt.Sprintf("Running `go mod vendor` in %q..", workingDirectory))
```
Comment (manicminer) [inline internal/tools/update-go-azure-sdk/main.go:524]:
```suggestion
	logger.Debug(fmt.Sprintf("Running `go mod vendor` in %q..", workingDirectory))
```

File: internal/tools/update-go-azure-sdk/main.go [hunk 2]
```diff
@@ -0,0 +1,729 @@
...
+	vendorArgs := []string{
+		"mod",
+		"vendor",
+	}
+	vendorCmd := exec.CommandContext(ctx, "go", vendorArgs...)
+	vendorCmd.Dir = workingDirectory
+	_ = vendorCmd.Start()
+	_ = vendorCmd.Wait()
+	logger.Debug(fmt.Sprintf("Run `go mod vendor` in %q.", workingDirectory))
```
Comment (manicminer) [inline internal/tools/update-go-azure-sdk/main.go:533]:
```suggestion
	logger.Debug(fmt.Sprintf("Running `go mod vendor` in %q.", workingDirectory))
```

File: internal/tools/update-go-azure-sdk/main.go [hunk 3]
```diff
@@ -0,0 +1,729 @@
...
+	if err != nil {
+		return fmt.Errorf("opening worktree: %+v", err)
+	}
+
+	// Whilst `Commit` has an add below it doesn't capture all changes, hence we need
+	// to explicitly stage all changes (including the vendored changes)
+	logger.Trace(fmt.Sprintf("Staging all changes in %q..", workingDirectory))
+	addOpts := &git.AddOptions{
+		All:  true,
+		Glob: "**",
```
Comment (manicminer) [inline internal/tools/update-go-azure-sdk/main.go:552]: I don't think `Glob` is needed with `All: true`? (might be faster to leave it out)

```suggestion
```

File: internal/tools/update-go-azure-sdk/main.go [hunk 4]
```diff
@@ -0,0 +1,729 @@
...
+	}
+	if err := worktree.AddWithOptions(addOpts); err != nil {
+		return fmt.Errorf("staging changes: %+v", err)
+	}
+
+	logger.Trace(fmt.Sprintf("Committing all changes in %q..", workingDirectory))
+	opts := &git.CommitOptions{
+		All: true,
```
Comment (manicminer) [inline internal/tools/update-go-azure-sdk/main.go:560]: Likewise, perhaps `All` not needed here if the index was already populated above?

File: internal/tools/update-go-azure-sdk/main.go [hunk 5]
```diff
@@ -0,0 +1,729 @@
...
+		fmt.Sprintf("github.com/hashicorp/go-azure-sdk/resource-manager/%s/", serviceName),
+		"--include=*.go",
+		"./internal/services/..",
+	}
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd := exec.CommandContext(ctx, "grep", diffArgs...)
+	cmd.Dir = absPath
+	cmd.Stderr = &stderr
+	cmd.Stdout = &stdout
```
Comment (manicminer) [inline internal/tools/update-go-azure-sdk/main.go:336]: Would it be more reliable / less forking to use `astutil.Imports()` for this instead?

**Review by manicminer — COMMENTED** [review]
@tombuildsstuff This looks great! Should we also run `make test` after updating the module version, to catch any breaking changes outside of the `services/` directory (e.g. `internal/client` or `internal/provider`)?

File: internal/tools/update-go-azure-sdk/main.go (hunk 3 above)
Comment (tombuildsstuff) [reply internal/tools/update-go-azure-sdk/main.go:552]: Tried that, but turns out the glob is also needed, else the vendor changes aren't picked up - so AFAIK this is required

File: internal/tools/update-go-azure-sdk/main.go (hunk 5 above)
Comment (tombuildsstuff) [reply internal/tools/update-go-azure-sdk/main.go:336]: Not a bad shout - will take a look.

**Comment by tombuildsstuff** [comment]
> Should we also run make test after updating the module version, to catch any breaking changes outside of the services/ directory (e.g. in internal/clients or internal/provider)?

I considered that, honestly if that's happened we have larger problems - either intentional or not - I suspect that we'll make those types of changes infrequently enough that it's not worth the cost of the extra CPU cycles to check that - since the tests would be run when the PR's sent to this repository anyway (where we'd catch it)?

**Comment by manicminer** [comment]
> I considered that, honestly if that's happened we have larger problems - either intentional or not - I suspect that we'll make those types of changes infrequently enough that it's not worth the cost of the extra CPU cycles to check that - since the tests would be run when the PR's sent to this repository anyway (where we'd catch it)?

I was mainly thinking of scenarios where an intentional breaking change is introduced, for example in `go-azure-sdk/sdk/environments`, where we intend to upgrade by hand - but thinking about it properly, the existing unit test workflow would of course be triggered by `vendor/**` anyway so I agree it's not worth the extra cycles 👍

**Pull request merged**